sql/src/test/java/io/crate/integrationtests/SQLTypeMappingTest.java

```diff
@@ -491,4 +491,43 @@ public void testThreeLevelNestedArrayColumn() throws Exception {
         assertThat((Integer)((Object[])((Object[])((Object[]) columns[0])[0])[0])[0], is(10));
         assertThat((Integer)((Object[])((Object[])((Object[]) columns[0])[0])[0])[1], is(20));
     }
+
+    @Test
+    public void testTwoLevelNestedObjectInArray() throws Exception {
+        execute("create table assets (\n" +
+                "    categories array(object (dynamic) as (\n" +
+                "       subcategories object (dynamic) as (\n" +
+                "          id int))))");
+        execute("insert into assets(categories) values ([{subcategories={id=10}}, {subcategories={id=20}}])");
+        ensureYellow();
+        execute("refresh table assets");
+
+        refresh();
+        waitNoPendingTasksOnAll();
+        SQLResponse response = execute("select categories['subcategories']['id'] from assets");
+        Object[] columns = TestingHelpers.getColumn(response.rows(), 0);
+        assertEquals(response.rowCount(), 1L);
+        assertThat((Integer) ((Object[]) columns[0])[0], is(10));
+        assertThat((Integer) ((Object[]) columns[0])[1], is(20));
+    }
+
+    @Test
+    public void testThreeLevelNestedObjectInArray() throws Exception {
```

Comment:
is the affected code really different between 2 or 3 nesting levels? if not remove this test..